File: Tayyibat/Services/SummaryService.swift
import Foundation
import SwiftData

/// يحدّث الملخصات اليومية اعتماداً على الوجبات المحفوظة.
enum SummaryService {
    /// يعيد حساب ملخص يوم معيّن من وجباته ويحفظه.
    @discardableResult
    static func updateSummary(for date: Date, context: ModelContext) -> DailySummary {
        let day = Calendar.current.startOfDay(for: date)
        let meals = mealsOfDay(day, context: context)
        let average = meals.isEmpty ? 0 : Int((Double(meals.map(\.overallScore).reduce(0, +)) / Double(meals.count)).rounded())

        let summary = summary(for: day, context: context) ?? {
            let new = DailySummary(date: day)
            context.insert(new)
            return new
        }()
        summary.averageScore = average
        summary.mealsCount = meals.count
        try? context.save()
        return summary
    }

    static func mealsOfDay(_ date: Date, context: ModelContext) -> [Meal] {
        let cal = Calendar.current
        let start = cal.startOfDay(for: date)
        guard let end = cal.date(byAdding: .day, value: 1, to: start) else { return [] }
        let predicate = #Predicate<Meal> { $0.capturedAt >= start && $0.capturedAt < end }
        let descriptor = FetchDescriptor<Meal>(predicate: predicate, sortBy: [SortDescriptor(\.capturedAt, order: .reverse)])
        return (try? context.fetch(descriptor)) ?? []
    }

    static func summary(for date: Date, context: ModelContext) -> DailySummary? {
        let day = Calendar.current.startOfDay(for: date)
        let predicate = #Predicate<DailySummary> { $0.date == day }
        let descriptor = FetchDescriptor<DailySummary>(predicate: predicate)
        return (try? context.fetch(descriptor))?.first
    }

    /// يضبط حالة الصيام لليوم في الملخص.
    static func setFasting(_ fasted: Bool, for date: Date, context: ModelContext) {
        let day = Calendar.current.startOfDay(for: date)
        let summary = summary(for: day, context: context) ?? {
            let new = DailySummary(date: day)
            context.insert(new)
            return new
        }()
        summary.fastedToday = fasted
        try? context.save()
    }

    /// عدد الأيام المتتالية (Streak) التي تجاوز متوسطها 80% حتى اليوم.
    static func currentStreak(context: ModelContext) -> Int {
        let cal = Calendar.current
        var streak = 0
        var day = cal.startOfDay(for: .now)
        while let summary = summary(for: day, context: context), summary.averageScore >= 80, summary.mealsCount > 0 {
            streak += 1
            guard let prev = cal.date(byAdding: .day, value: -1, to: day) else { break }
            day = prev
        }
        return streak
    }
}

File: android/app/src/main/java/com/tayyibat/app/service/SummaryService.kt
package com.tayyibat.app.service

import com.tayyibat.app.data.db.MealDao
import com.tayyibat.app.data.db.SummaryDao
import com.tayyibat.app.data.model.DailySummary
import com.tayyibat.app.data.model.MealWithItems
import com.tayyibat.app.util.DateUtils
import kotlin.math.roundToInt

/** يحدّث الملخصات اليومية اعتماداً على الوجبات المحفوظة. */
object SummaryService {

    /** يعيد حساب ملخص يوم معيّن من وجباته ويحفظه. */
    suspend fun updateSummary(forDayMillis: Long, mealDao: MealDao, summaryDao: SummaryDao): DailySummary {
        val dayStart = DateUtils.startOfDayMillis(forDayMillis)
        val meals = mealsOfDay(forDayMillis, mealDao)
        val average = if (meals.isEmpty()) 0
        else (meals.sumOf { it.meal.overallScore }.toDouble() / meals.size).roundToInt()

        val existing = summaryDao.summaryForDay(dayStart)
        val summary = (existing ?: DailySummary(date = dayStart)).copy(
            averageScore = average,
            mealsCount = meals.size,
        )
        summaryDao.upsert(summary)
        return summary
    }

    suspend fun mealsOfDay(dayMillis: Long, mealDao: MealDao): List<MealWithItems> {
        val start = DateUtils.startOfDayMillis(dayMillis)
        val end = start + 24L * 60 * 60 * 1000
        return mealDao.mealsBetween(start, end)
    }

    suspend fun summaryForDay(dayMillis: Long, summaryDao: SummaryDao): DailySummary? =
        summaryDao.summaryForDay(DateUtils.startOfDayMillis(dayMillis))

    /** يضبط حالة الصيام لليوم في الملخص. */
    suspend fun setFasting(fasted: Boolean, dayMillis: Long, summaryDao: SummaryDao) {
        val dayStart = DateUtils.startOfDayMillis(dayMillis)
        val existing = summaryDao.summaryForDay(dayStart) ?: DailySummary(date = dayStart)
        summaryDao.upsert(existing.copy(fastedToday = fasted))
    }

    /** عدد الأيام المتتالية (Streak) التي تجاوز متوسطها 80% حتى اليوم. */
    suspend fun currentStreak(summaryDao: SummaryDao): Int {
        var streak = 0
        var dayMillis = DateUtils.todayStartMillis()
        while (true) {
            val summary = summaryDao.summaryForDay(DateUtils.startOfDayMillis(dayMillis)) ?: break
            if (summary.averageScore >= 80 && summary.mealsCount > 0) {
                streak += 1
                dayMillis -= 24L * 60 * 60 * 1000
            } else break
        }
        return streak
    }
}
